Error: offset is longer than source length!  is thrown when I try to update inline snapshot

### Describe the bug

```
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

Error: offset is longer than source length! offset 947 > length 278
    at Object.numberToPos (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/source-map-f7a4da05.js:3390:11)
    at prepareSnapString (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:700:44)
    at replaceObjectSnap (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:696:59)
    at replaceInlineSnap (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:715:12)
    at /home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:679:7
    at async Promise.all (index 0)
    at async saveInlineSnapshots (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:671:3)
    at async SnapshotState.save (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:839:9)
    at async packSnapshotState (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:1007:18)
    at async SnapshotClient.saveSnap (/home/capaj/work-repos/mindmed/node-comparison/node_modules/vitest/dist/entry.js:986:20)
```

### Reproduction

I have this spec file
```ts
import { beforeAll, describe, expect, it } from 'vitest'
import { request } from 'undici'
import prismaClient from '../prisma/prismaClient'

describe('suite', () => {
  beforeAll(async () => {
    await prismaClient.questionnaire.deleteMany()
    await (
      await import('../server')
    ).listenPromise
  })

  describe('questionnares', () => {
    it('should GET', async () => {
      const res = await request('http://localhost:9090/questionnaires')

      expect(await res.body.json()).toMatchInlineSnapshot('[]')
    })

    it('should POST', async () => {
      const res = await request('http://localhost:9090/questionnaires', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          accept: 'application/json'
        },
        body: JSON.stringify({
          name: 'test',
          description: 'test',
          questionnaire: []
        })
      })

      expect(await res.body.json()).toMatchInlineSnapshot(
        {
          id: expect.any(Number)
        },
        `
          Object {
            "description": "test",
            "id": Any<Number>,
            "name": "test",
            "questionnaire": [],
          }
        `
      )
    })
  })
})

```

when I run the tests, I get this:
```
AssertionError: Snapshot `suite > questionnares > should POST 1` mismatched
 ❯ api/apiRoutes.spec.ts:34:36
     32|       })
     33| 
     34|       expect(await res.body.json()).toMatchInlineSnapshot(
       |                                    ^
     35|         {
     36|           id: expect.any(Number)

  - Expected  - 1
  + Received  + 1
  
  - Object {
  + {
      "description": "test",
      "id": Any<Number>,
      "name": "test",
      "questionnaire": [],

```

then I just hit `u` on the keyboard, error is thrown.

### System Info

```shell
System:
    OS: Linux 5.13 Ubuntu 21.10 21.10 (Impish Indri)
    CPU: (24) x64 AMD Ryzen 9 5900X 12-Core Processor
    Memory: 7.60 GB / 62.75 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 16.13.1 - ~/.nvm/versions/node/v16.13.1/bin/node
    Yarn: 1.22.17 - ~/.yarn/bin/yarn
    npm: 8.3.0 - ~/.nvm/versions/node/v16.13.1/bin/npm
  Browsers:
    Brave Browser: 97.1.34.80
    Chrome: 97.0.4692.99
    Firefox: 96.0
  npmPackages:
    vitest: ^0.2.1 => 0.2.1
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/snapshot/src/port/inlineSnapshot.ts
+++ b/packages/snapshot/src/port/inlineSnapshot.ts
@@ -34,19 +34,45 @@ export async function saveInlineSnapshots(
 const startObjectRegex = /(?:toMatchInlineSnapshot|toThrowErrorMatchingInlineSnapshot)\s*\(\s*(?:\/\*[\S\s]*\*\/\s*|\/\/.*\s+)*\s*({)/m
 
 function replaceObjectSnap(code: string, s: MagicString, index: number, newSnap: string) {
-  code = code.slice(index)
-  const startMatch = startObjectRegex.exec(code)
+  let _code = code.slice(index)
+  const startMatch = startObjectRegex.exec(_code)
   if (!startMatch)
     return false
 
-  code = code.slice(startMatch.index)
-  const charIndex = getCallLastIndex(code)
-  if (charIndex === null)
-    return false
+  _code = _code.slice(startMatch.index)
 
-  s.appendLeft(index + startMatch.index + charIndex, `, ${prepareSnapString(newSnap, code, index)}`)
+  let callEnd = getCallLastIndex(_code)
+  if (callEnd === null)
+    return false
+  callEnd += index + startMatch.index
+
+  const shapeStart = index + startMatch.index + startMatch[0].length
+  const shapeEnd = getObjectShapeEndIndex(code, shapeStart)
+  const snap = `, ${prepareSnapString(newSnap, code, index)}`
+
+  if (shapeEnd === callEnd) {
+    // toMatchInlineSnapshot({ foo: expect.any(String) })
+    s.appendLeft(callEnd, snap)
+  }
+  else {
+    // toMatchInlineSnapshot({ foo: expect.any(String) }, ``)
+    s.overwrite(shapeEnd, callEnd, snap)
+  }
 
   return true
+}
+
+function getObjectShapeEndIndex(code: string, index: number) {
+  let startBraces = 1
+  let endBraces = 0
+  while (startBraces !== endBraces && index < code.length) {
+    const s = code[index++]
+    if (s === '{')
+      startBraces++
+    else if (s === '}')
+      endBraces++
+  }
+  return index
 }
 
 function prepareSnapString(snap: string, source: string, index: number) {

--- a/test/core/test/inline-snap.test.ts
+++ b/test/core/test/inline-snap.test.ts
@@ -146,4 +146,44 @@ ${indent}}\`)
         "
     `)
   })
+
+  describe('replaceObjectSnap()', () => {
+    it('without snapshot', async () => {
+      const code = 'expect({ foo: \'bar\' }).toMatchInlineSnapshot({ foo: expect.any(String) })'
+
+      const s = new MagicString(code)
+      replaceInlineSnap(code, s, 23, `
+      {
+        "foo": Any<String>,
+      }
+    `)
+
+      expect(s.toString()).toMatchInlineSnapshot(`
+        "expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) }, \`
+          {
+                  \\"foo\\": Any<String>,
+                }
+        \`)"
+      `)
+    })
+
+    it('with snapshot', async () => {
+      const code = 'expect({ foo: \'bar\' }).toMatchInlineSnapshot({ foo: expect.any(String) }, `{ }`)'
+
+      const s = new MagicString(code)
+      replaceInlineSnap(code, s, 23, `
+      {
+        "foo": Any<String>,
+      }
+    `)
+
+      expect(s.toString()).toMatchInlineSnapshot(`
+        "expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) }, \`
+          {
+                  \\"foo\\": Any<String>,
+                }
+        \`)"
+      `)
+    })
+  })
 })

--- a/test/snapshots/test-update/snapshots-inline-js.test.js
+++ b/test/snapshots/test-update/snapshots-inline-js.test.js
@@ -23,3 +23,21 @@ describe('snapshots are generated in correct order', async () => {
     `)
   })
 })
+
+describe('snapshots with properties', () => {
+  test('without snapshot', () => {
+    expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) }, `
+      Object {
+        "foo": Any<String>,
+      }
+    `)
+  })
+
+  test('with snapshot', () => {
+    expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) }, `
+      Object {
+        "foo": Any<String>,
+      }
+    `)
+  })
+})

--- a/test/snapshots/tools/inline-test-template.js
+++ b/test/snapshots/tools/inline-test-template.js
@@ -11,3 +11,13 @@ describe('snapshots are generated in correct order', async () => {
     expect({ foo: ['zed'] }).toMatchInlineSnapshot()
   })
 })
+
+describe('snapshots with properties', () => {
+  test('without snapshot', () => {
+    expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) })
+  })
+
+  test('with snapshot', () => {
+    expect({ foo: 'bar' }).toMatchInlineSnapshot({ foo: expect.any(String) }, '')
+  })
+})

